### DISCUSSION TITLE
Failed installation test

### DISCUSSION THREAD

**Discussion by Nickfu0911** [firedrakeproject/firedrake#3512] [Firedrake support]
Hello,

I successfully installed Firedrake and tried to run the installation tests. But it failed. 
In the terminal it shows: 

======================================================================================== test session starts =========================================================================================
platform darwin -- Python 3.11.9, pytest-8.1.1, pluggy-1.4.0
rootdir: /Users/nf419/firedrake/src/firedrake
configfile: setup.cfg
plugins: anyio-4.3.0, mpi-0.1, xdist-3.5.0, nbval-0.11.0
collected 3517 items / 3490 deselected / 1 skipped / 27 selected                                                                                                                                     

tests/regression/test_dg_advection.py .F.F                                                                                                                                                     [ 14%]
tests/regression/test_poisson_strong_bcs.py ................F                                                                                                                                  [ 77%]
tests/regression/test_poisson_strong_bcs_nitsche.py ....                                                                                                                                       [ 92%]
tests/regression/test_stokes_mini.py ..                                                                                                                                                        [100%]

============================================================================================== FAILURES ==============================================================================================
___________________________________________________________________________ test_dg_advection_icosahedral_sphere_parallel ____________________________________________________________________________

args = (), kwargs = {}

    def parallel_callback(*args, **kwargs):
>       subprocess.run(cmd, check=True)

../pytest-mpi/pytest_mpi.py:192: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

input = None, capture_output = False, timeout = None, check = True, popenargs = (['mpiexec', '-n', '1', '-genv', '_PYTEST_MPI_CHILD_PROCESS', '1', ...],), kwargs = {}
process = <Popen: returncode: 59 args: ['mpiexec', '-n', '1', '-genv', '_PYTEST_MPI_CH...>, stdout = None, stderr = None, retcode = 59

    def run(*popenargs,
            input=None, capture_output=False, timeout=None, check=False, **kwargs):
        """Run command with arguments and return a CompletedProcess instance.
    
        The returned instance will have attributes args, returncode, stdout and
        stderr. By default, stdout and stderr are not captured, and those attributes
        will be None. Pass stdout=PIPE and/or stderr=PIPE in order to capture them,
        or pass capture_output=True to capture both.
    
        If check is True and the exit code was non-zero, it raises a
        CalledProcessError. The CalledProcessError object will have the return code
        in the returncode attribute, and output & stderr attributes if those streams
        were captured.
    
        If timeout is given, and the process takes too long, a TimeoutExpired
        exception will be raised.
    
        There is an optional argument "input", allowing you to
        pass bytes or a string to the subprocess's stdin.  If you use this argument
        you may not also use the Popen constructor's "stdin" argument, as
        it will be used internally.
    
        By default, all communication is in bytes, and therefore any "input" should
        be bytes, and the stdout and stderr will be bytes. If in text mode, any
        "input" should be a string, and stdout and stderr will be strings decoded
        according to locale encoding, or by "encoding" if set. Text mode is
        triggered by setting any of text, encoding, errors or universal_newlines.
    
        The other arguments are the same as for the Popen constructor.
        """
        if input is not None:
            if kwargs.get('stdin') is not None:
                raise ValueError('stdin and input arguments may not both be used.')
            kwargs['stdin'] = PIPE
    
        if capture_output:
            if kwargs.get('stdout') is not None or kwargs.get('stderr') is not None:
                raise ValueError('stdout and stderr arguments may not be used '
                                 'with capture_output.')
            kwargs['stdout'] = PIPE
            kwargs['stderr'] = PIPE
    
        with Popen(*popenargs, **kwargs) as process:
            try:
                stdout, stderr = process.communicate(input, timeout=timeout)
            except TimeoutExpired as exc:
                process.kill()
                if _mswindows:
                    # Windows accumulates the output in a single blocking
                    # read() call run on child threads, with the timeout
                    # being done in a join() on those threads.  communicate()
                    # _after_ kill() is required to collect that and add it
                    # to the exception.
                    exc.stdout, exc.stderr = process.communicate()
                else:
                    # POSIX _communicate already populated the output so
                    # far into the TimeoutExpired exception.
                    process.wait()
                raise
            except:  # Including KeyboardInterrupt, communicate handled that.
                process.kill()
                # We don't call process.wait() as .__exit__ does that for us.
                raise
            retcode = process.poll()
            if check and retcode:
>               raise CalledProcessError(retcode, process.args,
                                         output=stdout, stderr=stderr)
E               subprocess.CalledProcessError: Command '['mpiexec', '-n', '1', '-genv', '_PYTEST_MPI_CHILD_PROCESS', '1', 'python', '-m', 'pytest', '--runxfail', '-s', '-q', '/Users/nf419/firedrake/src/firedrake/tests/regression/test_dg_advection.py::test_dg_advection_icosahedral_sphere_parallel', ':', '-n', '2', 'python', '-m', 'pytest', '--runxfail', '-s', '-q', '/Users/nf419/firedrake/src/firedrake/tests/regression/test_dg_advection.py::test_dg_advection_icosahedral_sphere_parallel', '--tb=no', '--no-summary', '--no-header', '--disable-warnings', '--show-capture=no']' returned non-zero exit status 59.

/opt/homebrew/Cellar/python@3.11/3.11.9/Frameworks/Python.framework/Versions/3.11/lib/python3.11/subprocess.py:571: CalledProcessError
---------------------------------------------------------------------------------------- Captured stderr call ----------------------------------------------------------------------------------------
firedrake:WARNING OMP_NUM_THREADS is not set or is set to a value greater than 1, we suggest setting OMP_NUM_THREADS=1 to improve performance
[0]PETSC ERROR: ------------------------------------------------------------------------
[0]PETSC ERROR: Caught signal number 11 SEGV: Segmentation Violation, probably memory access out of range
[0]PETSC ERROR: Try option -start_in_debugger or -on_error_attach_debugger
[0]PETSC ERROR: or see https://petsc.org/release/faq/#valgrind and https://petsc.org/release/faq/
[0]PETSC ERROR: configure using --with-debugging=yes, recompile, link, and run 
[0]PETSC ERROR: to get more information on the crash.
[0]PETSC ERROR: Run with -malloc_debug to check if memory corruption is causing the crash.
application called MPI_Abort(MPI_COMM_WORLD, 59) - process 0

And very similar error messages for the other tests. 
May I ask what's the main problem here? 

I also tried to run PETSC make check, and it gives: 

Running PETSc check examples to verify correct installation
Using PETSC_DIR=/Users/nf419/firedrake/src/petsc and PETSC_ARCH=default
C/C++ example src/snes/tutorials/ex19 run successfully with 1 MPI process
C/C++ example src/snes/tutorials/ex19 run successfully with 2 MPI processes
C/C++ example src/snes/tutorials/ex19 run successfully with HYPRE
3c3
<     0 KSP Residual norm 0.235858
---
>     0 KSP Residual norm 0.0348072
5,9c5,52
<   1 SNES Function norm 6.81968e-05
<     0 KSP Residual norm 2.30906e-05
<     1 KSP Residual norm < 1.e-11
<   2 SNES Function norm < 1.e-11
< Number of SNES iterations = 2
---
>   1 SNES Function norm 0.211774
> [0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
> [0]PETSC ERROR: Object is in wrong state
> [0]PETSC ERROR: Outstanding operation has not been completed
> [0]PETSC ERROR: WARNING! There are unused option(s) set! Could be the program crashed before usage or a spelling mistake, etc!
> [0]PETSC ERROR:   Option left: name:-nox_warning (no value) source: environment
> [0]PETSC ERROR: See https://petsc.org/release/faq/ for trouble shooting.
> [0]PETSC ERROR: Petsc Development GIT revision: v3.4.2-44465-g272f13d92c7  GIT Date: 2024-03-22 13:06:44 +0000
> [0]PETSC ERROR: ./ex19 on a default named IC-GTDG92R571 by nf419 Thu Apr 18 19:48:52 2024
> [0]PETSC ERROR: Configure options PETSC_DIR=/Users/nf419/firedrake/src/petsc PETSC_ARCH=default --download-chaco --download-hwloc-configure-arguments=--disable-opencl --download-netcdf --with-fortran-bindings=0 --download-bison --download-hdf5 --with-c2html=0 --download-mpich-configure-arguments=--disable-opencl --download-openblas-make-options="'USE_THREAD=0 USE_LOCKING=1 USE_OPENMP=0'" --with-zlib --CFLAGS="-Wno-implicit-function-declaration -Wno-int-conversion -Wno-implicit-int" --with-debugging=0 --download-pastix --LDFLAGS=-Wl,-ld_classic --download-suitesparse --download-scalapack --download-metis --download-superlu_dist --download-pnetcdf --with-x=0 --download-mpich --download-mumps --with-shared-libraries=1 --download-hypre --download-hwloc --download-ptscotch --download-openblas
> [0]PETSC ERROR: #1 PetscSFReset_Basic() at /Users/nf419/firedrake/src/petsc/src/vec/is/sf/impls/basic/sfbasic.c:261
> [0]PETSC ERROR: #2 PetscSFReset() at /Users/nf419/firedrake/src/petsc/src/vec/is/sf/interface/sf.c:117
> [0]PETSC ERROR: #3 PetscSFDestroy() at /Users/nf419/firedrake/src/petsc/src/vec/is/sf/interface/sf.c:239
> [0]PETSC ERROR: #4 VecScatterDestroy() at /Users/nf419/firedrake/src/petsc/src/vec/is/sf/interface/vscat.c:480
> [0]PETSC ERROR: #5 MatFactorNumeric_MUMPS() at /Users/nf419/firedrake/src/petsc/src/mat/impls/aij/mpi/mumps/mumps.c:2056
> [0]PETSC ERROR: #6 MatLUFactorNumeric() at /Users/nf419/firedrake/src/petsc/src/mat/interface/matrix.c:3253
> [0]PETSC ERROR: #7 PCSetUp_LU() at /Users/nf419/firedrake/src/petsc/src/ksp/pc/impls/factor/lu/lu.c:121
> [0]PETSC ERROR: #8 PCSetUp() at /Users/nf419/firedrake/src/petsc/src/ksp/pc/interface/precon.c:1079
> [0]PETSC ERROR: #9 KSPSetUp() at /Users/nf419/firedrake/src/petsc/src/ksp/ksp/interface/itfunc.c:415
> [0]PETSC ERROR: #10 KSPSolve_Private() at /Users/nf419/firedrake/src/petsc/src/ksp/ksp/interface/itfunc.c:831
> [0]PETSC ERROR: #11 KSPSolve() at /Users/nf419/firedrake/src/petsc/src/ksp/ksp/interface/itfunc.c:1078
> [0]PETSC ERROR: #12 PCApply_FieldSplit_Schur() at /Users/nf419/firedrake/src/petsc/src/ksp/pc/impls/fieldsplit/fieldsplit.c:1187
> [0]PETSC ERROR: #13 PCApply() at /Users/nf419/firedrake/src/petsc/src/ksp/pc/interface/precon.c:497
> [0]PETSC ERROR: #14 KSP_PCApply() at /Users/nf419/firedrake/src/petsc/include/petsc/private/kspimpl.h:409
> [0]PETSC ERROR: #15 KSPInitialResidual() at /Users/nf419/firedrake/src/petsc/src/ksp/ksp/interface/itres.c:64
> [0]PETSC ERROR: #16 KSPSolve_GMRES() at /Users/nf419/firedrake/src/petsc/src/ksp/ksp/impls/gmres/gmres.c:226
> [0]PETSC ERROR: #17 KSPSolve_Private() at /Users/nf419/firedrake/src/petsc/src/ksp/ksp/interface/itfunc.c:905
> [0]PETSC ERROR: #18 KSPSolve() at /Users/nf419/firedrake/src/petsc/src/ksp/ksp/interface/itfunc.c:1078
> [0]PETSC ERROR: #19 SNESSolve_NEWTONLS() at /Users/nf419/firedrake/src/petsc/src/snes/impls/ls/ls.c:220
> [0]PETSC ERROR: #20 SNESSolve() at /Users/nf419/firedrake/src/petsc/src/snes/interface/snes.c:4738
> [0]PETSC ERROR: #21 main() at ex19.c:152
> [0]PETSC ERROR: PETSc Option Table entries:
> [0]PETSC ERROR: -checkfunctionlist (source: environment)
> [0]PETSC ERROR: -fieldsplit_0_pc_factor_mat_solver_type mumps (source: command line)
> [0]PETSC ERROR: -fieldsplit_0_pc_type lu (source: command line)
> [0]PETSC ERROR: -fieldsplit_1_pc_factor_mat_solver_type mumps (source: command line)
> [0]PETSC ERROR: -fieldsplit_1_pc_type lu (source: command line)
> [0]PETSC ERROR: -ksp_monitor_short (source: command line)
> [0]PETSC ERROR: -nox_warning (source: environment)
> [0]PETSC ERROR: -pc_fieldsplit_0_fields 0,1,2 (source: command line)
> [0]PETSC ERROR: -pc_fieldsplit_1_fields 3 (source: command line)
> [0]PETSC ERROR: -pc_fieldsplit_block_size 4 (source: command line)
> [0]PETSC ERROR: -pc_fieldsplit_type SCHUR (source: command line)
> [0]PETSC ERROR: -pc_type fieldsplit (source: command line)
> [0]PETSC ERROR: -snes_monitor_short (source: command line)
> [0]PETSC ERROR: -use_gpu_aware_mpi 0 (source: environment)
> [0]PETSC ERROR: ----------------End of Error Message -------send entire error message to petsc-maint@mcs.anl.gov----------
> application called MPI_Abort(MPI_COMM_SELF, 73) - process 0
/Users/nf419/firedrake/src/petsc/src/snes/tutorials
Possible problem with ex19 running with MUMPS, diffs above
=========================================
C/C++ example src/snes/tutorials/ex19 run successfully with SuiteSparse
C/C++ example src/snes/tutorials/ex19 run successfully with SuperLU_DIST
C/C++ example src/vec/vec/tests/ex47 run successfully with HDF5
Completed PETSc check examples
Error while running make check
make[1]: *** [check] Error 1
make: *** [check] Error 2

**Comment by Ig-dolci** [comment]
The issue causing errors in executing a Firedrake code on a Mac in parallel is related to [PETSc issue #1569](https://gitlab.com/petsc/petsc/-/issues/1569). While serial Firedrake simulations work, parallel execution is currently problematic. If you need to run codes in parallel, a workaround is to use a Firedrake Docker image. Otherwise, your installation should handle serial codes perfectly.

**Reply by Nickfu0911** [reply]
Hi,

Thanks for the reply. So if I'm only going to use serial codes, then I don't need to worry about this error in the installation test right? 

Thanks

**Reply by Ig-dolci** [reply]
Yes, the serial code should work without issues.
